auth/net_tools.go

```diff
@@ -40,27 +40,34 @@ func getIPAddress(r *http.Request) string {
 	return r.RemoteAddr
 }
 
-func auditUserLogAction(central *ImqsCentral, req *httpRequest, userId authaus.UserId, username, description string, actionType authaus.AuditActionType) {
+func auditUserLogAction(central *ImqsCentral, req *httpRequest, contextUserId authaus.UserId, contextUsername, description string, actionType authaus.AuditActionType) {
 	var actorUserId authaus.UserId
 
-	var email string
-	user, eUser := central.Central.GetUserFromIdentity(username)
+	var contextEmail string
+	var contextUser authaus.AuthUser
+	var eUser error
+	if contextUsername != "" {
+		contextUser, eUser = central.Central.GetUserFromIdentity(contextUsername)
+	} else {
+		contextUser, eUser = central.Central.GetUserFromUserId(contextUserId)
+	}
+
 	if eUser == nil {
-		email = user.Email
+		contextEmail = contextUser.Email
 	}
 
 	contextDetails := ContextDetails{
 		Service:  "auth",
 		Origin:   getIPAddress(req.http),
-		Username: username,
-		UserId:   int64(userId),
-		Email:    email,
+		Username: contextUsername,
```

Comment:
Would `contextUser.Username` still be an empty string if `contextUsername` was an empty string? If not, could you please change this to rather use `contextUser.Username`

Reply:
This is what i meant when i said that the code is a mess. As far as your question goes, it depends how the function is called in the first place. It is possible that for some reason the caller included a `contextUsername` that is not linked to an actual user in the auth db...in which case we need to decide if it needs to go into the 'context' portion of the audit log or if it sufficient to just have it in the 'who' column. If you check lower down, the 'who' coloumn _also_ has some dependency on the `actorUserId` so even that is not certain. My point is that we could end up with a situation where the contextUserName supplied is not recorded anywhere in the audit trail (since there are `getIdentity..` calls involved which may return a different value...which is probably not what we want. This whole thing needs a refactor because the intention of the function is not clear at all - i think it got bastardised over the years to mean different things in different contexts. Thoughts?

Reply:
...also, we should not forget that the front-end also assumes certain things when populating the dropdown for audit log filtering which complicates things here.

Reply:
I may have just opened a small can of worms...

So I was looking through all the references to this function. And everywhere they use the username associated with the userid, so my suggestion is to change this function to take a `AuthUser` instead of the `userId` and `username` parameters.

But now for the can of worms... Almost all calls to this function is wrong... They all use the the username and ID of the **affected user**... not the user that actually did the action :(.

Reply:
Yes...which means in some cases the "context" is treated as the "target".

Reply:
TL;DR: So the answer to your question: I tried to leave the function's behaviour as much the same as possible, which includes leaving variables exactly as they were used before, just focusing on eliminating the very wrong email address that was included in the context object. Barring some variable name changes of course. Keep in mind this was only a problem for Group changes etc where the caller and the context object details should match. As you pointed out all other calls includes the target user's details in the context object.

More detail:
Just to get back to the original reason for this change. There are users in the DOI environments which for some reason do not have usernames (""), but they do have emails, which allows them to log in. As for the audit log, the missing usernames caused multiple records to be returned inside one of the "find" functions (because the function searches for an email OR username) and then it just takes the first one that matches - resulting in the wrong email address being included in the context details. This fix prevents a blank username to be used for identification, otherwise it would just operate as before. Which would still leave us with a blank username, but at least have the email address match the user id in the context object. 
So with that view, for now i would vote for not using `user.Username` (since it would be blank anyway in the context where this was an issue (group permission updates), in other contexts it could change the symantics...it would depend what is passed in as userid in those cases. I'm not sure.

As it stands, a user without username that changes e.g. another users details (telephone nr), would not be traceable via the audit log, since in those cases, the _target_ user's details are added into the context (obviously wrong, but that's how it's been for quite a while) and the username is blank so the _who_ field is also blank. It went wrong around here `e682ccfe` when the username and id parameters were added to the call...anyways. This was 5+ years ago.